### PR TITLE
build into directories based on flags

### PR DESCRIPTION
**Pull request by rtjohnso**
This PR attempts to make builds more reliable, even as we change things like `CFLAGS`, `LDFLAGS`, `CC`, etc.

All builds go into `build/<hash>`, where `<hash>` is a hash of all the relevant make variables, such as `CFLAGS`, etc.

It then creates symlinks `bin`, `obj`, and `lib` in the top-level directory.

Unfortunately, this doesn't work with target-specific variables, so I had to get rid of `make release`, `make debug`, and `make debug-log`.  Instead, you can do 
`make D=<anything>`
or
`make DL=<anything>`
for debug or debug-log mode, respectively.

For debug builds, the top-level directories are `bin-debug`, `lib-debug`, and `obj-debug`.

The makefile also saves all relevant make variables in `build/<hash>/flags`.

`make clean` blows away the whole `build` directory.

### CODE REVIEW

File: Makefile
Comment (gapisback) [inline Makefile:89]: ? Delete this line, if we don't need it any more ?

File: Makefile
Comment (gapisback) [inline Makefile:137]: Seems like this can be merged with prev line using `else ifneq "$(BUILD_MSAN)" "0"` --- for readability.

File: Makefile
Comment (gapisback) [inline Makefile:117]: (Saw this one later ...) Seems like this can be changed to use `else ifneq "$(BUILD_ASAN)" "0"`

File: Makefile
Comment (gapisback) [inline Makefile:235]: As these `Makefile` symbols are essentially derived from  any existing `BUILD_*` env-var settings, can you also consider dumping the following to the config check file:

`env | grep BUILD_ >> $@`

Add this to before L235.

File: Makefile
Comment (gapisback) [inline Makefile:88]: For readability, it will be better to indent the block under if / else / elseif. I tried with an indent of 3-spaces, and the scripts still seemed to work.

If you can do this indentation in all places where the if/elseif blocks are used, in this pass, that will be great. 

Otherwise, feel free to push this to a follow-on issue. (I'm collecting up some mop-up items to be done after this change merges to main.)

File: Makefile
Comment (rtjohnso) [reply Makefile:230]: That would not be accurate.  Even if the config file matches, we may need to build some stuff (e.g. if you edited one of the source files).

**Review by gapisback — COMMENTED** [review]
Hi, @rtjohnso - I am [finally] passing this review. Thanks for making this happen!

I applied a set of diffs on top of this branch in my private dev branch, [robj/make-dirs-apg-fix-CI](https://github.com/vmware/splinterdb/tree/robj/make-dirs-apg-fix-CI), under this commit:
https://github.com/vmware/splinterdb/pull/369/commits/4c76aba43cd37a6ad5483c947074e84277bdeb00

Please see my description in PR #369 explaining how you could integrate those changes on top of this work.

I did not pay attention to @ajhconway 's issues of CC-flags in diff places; too low-level for me in this round. We may have to make a 2nd pass to ensure that the right flags have been used in all places.

**Review by gapisback — APPROVED** [review]
Approved, once we can pull-in changes from  the last commit in branch [robj/make-dirs-apg-fix-CI](https://github.com/vmware/splinterdb/tree/robj/make-dirs-apg-fix-CI) to this branch.

**Comment by gapisback** [comment]
@rtjohnso - I am closing my efforts on this `Makefile` PR. Just wanted to remark that your original description is way out-of-sync with the final implementation. And there are messy multiple commits that have come in.

[Unsolicited suggestion] Here is a workflow I would follow to move this forward:

1. Cherry-pick and adjust the diffs from the one commit I provided separately
2. Squash all your commits (howsoever you can do that) to one commit
3. Rewrite that single commit's message using `git commit --amend` and describe the functionality and interface (`BUILD_` env-vars) for the new `Makefile` behaviour
4. You would have to rebase v/s /main as I merged two required commits to /main earlier today.
5. `git push --force` to your branch, which will re-fire up CI jobs. I expect they will all pass
6. Then rebase push into /main.

If you want me to give a final sanity re-look at the final-final merged-commit, I'd be happy to do so. Let me know when you might be ready.

I'm sure you are aware of these dev-processes, but just pointing them out here, mainly to ensure that 

- (a) we get a single commit for this work in /main, and 
- (b) we get a well-described commit message which will act like a quick-read FAQ / how-to to use this nifty interface.

Thanks for all your hard working and plumbing this through e2e!

**Comment by rtjohnso** [comment]
OMG I had a dumb bug in make help.  Sorry about that.

As for the ar failure, can you send me the output of make VERBOSE=1?

As for the CI, it fails because this PR changes the interface to the
Makefile.  We will need to update the CI.

Best,
Rob

**Pull request merged**
